crypto/src/symkey/chacha20poly1305/mod.rs

```diff
@@ -0,0 +1,308 @@
+        key: Vec<u8>,
+        ad: Vec<u8>,
+        nonce: Vec<u8>,
+        input: Vec<u8>,
+        output: Vec<u8>,
+    }
+
+    impl ExternalXChacha20Poly1305Data {
+        fn from_map(map: BTreeMap<String, Vec<u8>>) -> Self {
+            Self {
+                key: map.get("Key").unwrap().clone(),
+                ad: map.get("AD").unwrap().clone(),
+                nonce: map.get("Nonce").unwrap().clone(),
+                input: map.get("In").unwrap().clone(),
+                output: map.get("Out").unwrap().clone(),
+            }
+        }
+    }
+
+    fn read_test_vectors_file<P: AsRef<std::path::Path>>(
+        path: P,
+    ) -> Vec<ExternalXChacha20Poly1305Data> {
+        let f = std::fs::File::open(path).unwrap();
+        let lines = io::BufReader::new(f).lines();
+
+        let mut result = Vec::new();
+
+        let mut current_batch_map: BTreeMap<String, Vec<u8>> = Default::default();
+        for (_line_num, line) in lines.into_iter().enumerate() {
+            let line = line.unwrap();
+            let line = line.trim();
+            if line.starts_with('#') {
+                continue;
+            }
+            // an empty line means we're starting a new test
+            if line.is_empty() {
+                if !current_batch_map.is_empty() {
+                    result.push(ExternalXChacha20Poly1305Data::from_map(current_batch_map));
+                }
+                current_batch_map = Default::default();
+                continue;
+            }
+
+            // the only remaining case is for "some_key = some_hex_value" in the line
+            let split_parts = line.split('=').collect::<Vec<_>>();
+            assert!(split_parts.len() == 2);
+
+            // insert this information to the map
+            let k = split_parts[0].trim();
+            let v = split_parts[1].trim();
+            let el = Vec::from_hex(v).unwrap();
+            current_batch_map.insert(k.to_owned(), el);
+        }
+        result
+    }
+
+    fn test_external_data(data: ExternalXChacha20Poly1305Data) {
+        let message = data.input;
+        let cipher = data.output;
+        let key_bin = data.key;
+        let aead = data.ad;
+        let nonce = data.nonce;
+        let key = Chacha20poly1305Key::new_from_array(key_bin.try_into().unwrap());
+        let nonce_with_encrypted =
+            key.encrypt_with_nonce_and_aead(&nonce, message.as_slice(), &aead).unwrap();
+        let decrypted = key
+            .decrypt_with_nonce_and_aead(&nonce, &nonce_with_encrypted[NONCE_LEN..], Some(&aead))
+            .unwrap();
+        assert_eq!(nonce_with_encrypted[NONCE_LEN..], cipher);
+        assert_eq!(decrypted, message);
+    }
+
+    #[test]
+    fn select_text_external() {
+        let test_vectors = read_test_vectors_file(
+            std::path::Path::new("src")
+                .join("symkey")
+                .join("chacha20poly1305")
+                .join("XCHACHA20_POLY1305_TEST_VECTORS_QIG_CRL_2019_10_17.tv"),
```

Comment:
I'm tempted to say we rename this to something take makes more sense for us. The file name won't mean a lot to most people, we could probably strip the header in the file too since it just refers to a bunch of people that nobody knows

Reply:
Yeah he more I think about it cleaning it up is probably worth the 10 mins it will take